packages/nx/src/plugins/js/project-graph/build-dependencies/target-project-locator.ts

```diff
@@ -1,45 +1,74 @@
-import { workspaceRoot } from '../../../../utils/workspace-root';
+import { readFileSync } from 'node:fs';
+import { builtinModules } from 'node:module';
+import { dirname, join, posix, relative } from 'node:path';
 import {
   ProjectGraphExternalNode,
   ProjectGraphProjectNode,
 } from '../../../../config/project-graph';
-import { builtinModules } from 'module';
 import {
   createProjectRootMappings,
   findProjectForPath,
 } from '../../../../project-graph/utils/find-project-for-path';
+import { isRelativePath, readJsonFile } from '../../../../utils/fileutils';
+import { parseJson } from '../../../../utils/json';
+import { workspaceRoot } from '../../../../utils/workspace-root';
+import { findExternalPackageJsonPath } from '../../utils/find-external-package-json-path';
+import {
+  getRootTsConfigFileName,
+  resolveModuleByImport,
+} from '../../utils/typescript';
 
-const builtInModuleSet = new Set<string>([
+/**
+ * The key is a combination of the package name and the project root importing it
+ * e.g. `lodash__packages/my-lib`, the value is the resolved external node name
+ * from the project graph.
+ */
+export type NpmResolutionCache = Map<string, string>;
+
+export const builtInModuleSet = new Set<string>([
   ...builtinModules,
   ...builtinModules.map((x) => `node:${x}`),
 ]);
 
+export function isBuiltinModuleImport(importExpr: string): boolean {
+  const packageName = parsePackageNameFromImportExpression(importExpr);
+  return builtInModuleSet.has(packageName);
+}
+
 export class TargetProjectLocator {
   private projectRootMappings = createProjectRootMappings(this.nodes);
-  private npmProjects = filterRootExternalDependencies(this.externalNodes);
+  private npmProjects = Object.keys(this.externalNodes)
+    .filter((k) => k.startsWith('npm:'))
+    .map((k) => this.externalNodes[k]);
   private tsConfig = this.getRootTsConfig();
   private paths = this.tsConfig.config?.compilerOptions?.paths;
   private typescriptResolutionCache = new Map<string, string | null>();
-  private npmResolutionCache = new Map<string, string | null>();
 
   constructor(
     private readonly nodes: Record<string, ProjectGraphProjectNode>,
-    private readonly externalNodes: Record<string, ProjectGraphExternalNode>
+    private readonly externalNodes: Record<
+      string,
+      ProjectGraphExternalNode
+    > = {},
+    private readonly npmResolutionCache: NpmResolutionCache = new Map<
+      string,
+      string
+    >()
   ) {}
 
   /**
-   * Find a project based on its import
+   * Resolve any workspace or external project that matches the given import expression,
+   * originating from the given filePath, relative to the given project root.
    *
    * @param importExpr
    * @param filePath
+   * @param projectRoot
    */
-  findProjectWithImport(importExpr: string, filePath: string): string {
+  findProjectFromImport(
+    importExpr: string,
+    filePath: string,
+    projectRoot: string
```

Comment:
I don't think this should be limited to project roots. It should be able to use the `dirname` of the `filePath`?